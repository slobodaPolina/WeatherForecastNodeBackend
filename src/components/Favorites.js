import React from 'react';
import Forecast from './Forecast';
import SmallPreview from './SmallPreview';
import Loading from './Loading';
import { getCityByName, getTheFavorites, addCityToTheFavorites, deleteCityFromTheFavorites } from "../cityAdder";

export class Favorites extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: "", // it is the text from the searchBar
            isCityValid: true, // is the city user tried t add valid
            favorites: [] // array of objects of favorite cities. We keep only names in db
        };
        this.handleSearch = this.handleSearch.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        getTheFavorites(
            (data) => {
                data.forEach(name => {
                    getCityByName(
                        name,
                        (data) => {
                            this.addToTheFavorites(Object.assign(data, { loaded: true }));
                        },
                        () => {
                            console.error("Error getting favorite city " + name + " from weather API");
                            this.addToTheFavorites({
                                name,
                                loaded: false
                            });
                        }
                    );
                });
            },
            () => console.error("Problems getting list of favorite cities from db!")
        );
    }

    render() {
        let msg = this.state.isCityValid ? "" : <div className="error-message">Seems like the city is not valid</div>;
        return (
            <div>
                <div>
                    <div className="search-bar row">
                        <div className="bigText">Избранное</div>
                        <div>
                            <form onSubmit={this.handleSearch}>
                                <input id="search-input" type="search"
                                       placeholder="Добавить новый город" value={this.state.city} onChange={this.handleInputChange}/>
                                <button className="search-button roundButton">+</button>
                            </form>
                        </div>
                    </div>
                    {msg}
                </div>
                <div className="favorites">
                    {this.state.favorites.map((city) => {
                        return (
                            <div className="smallCityCard" key={city.name}>
                                <SmallPreview city={city} removeCity={this.removeGenerator(city.name)}/>
                                {city.loaded ? <Forecast data={city}/> : <Loading />}
                            </div>
                        )
                    })}
                </div>
            </div>
        );
    }

    handleInputChange(event) {
        this.setState({city: event.target.value});
    }

    handleSearch(event) {
        const getCallback = (isSuccess) => () => {
            this.setState({isCityValid: isSuccess});
            // todo remove loading card anyway
            if (isSuccess) { // means we got data from weather service
                addCityToTheFavorites(
                    this.state.city,
                    () => {
                        this.setState({favorites: this.state.favorites.push(this.state.city)});
                        this.setState({city: ""});
                    },
                    () => {
                        console.error('error during saving favorite city in db!');
                        this.setState({isCityValid: false});
                    }
                );
            }
        };

        event.preventDefault();
        if (this.props.favorites.find(city => city.name === this.state.city)) { // if we already have it we wont add it again
            this.setState({isCityValid: false});
        } else {
            // todo start the loader on the citycard
            // load real data and display it
            getCityByName(
                this.state.city,
                getCallback(true),
                getCallback(false)
            );
        }
    }

    removeGenerator(cityName) {
        return () => deleteCityFromTheFavorites(
            cityName,
            () => {
                this.setState({
                    favorites: this.state.favorites.filter(city => city.name !== cityName)
                })
            },
            () => { console.error("Problems removing city " + cityName + " from db!") }
         );
    }

    addToTheFavorites(obj) {
        this.setState({
            favorites: this.state.favorites.push(obj)
        });
    }
}

export default Favorites;
