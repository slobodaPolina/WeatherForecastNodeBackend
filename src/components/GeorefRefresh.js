import React from 'react';
import { connect } from 'react-redux';
import { addCityByGeolocation } from '../cityAdder';


export class GeorefRefresh extends React.Component {
    constructor(props) {
        super(props);
        this.addCity = this.addCity.bind(this);
    }

    addCity() {
        this.props.dispatch(addCityByGeolocation());
    }

    render() {
        return (
            <button className="georefRefresh"
                onClick={this.addCity}>
                    Обновить геолокацию
            </button>
        );
    }
}

export default connect()(GeorefRefresh);
