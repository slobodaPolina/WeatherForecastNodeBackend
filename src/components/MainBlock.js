import React from 'react';
import { connect } from 'react-redux';
import { addCityByGeolocation } from '../cityAdder';
import MainPreview from './MainPreview';
import Forecast from './Forecast';
import Loading from './Loading';

export class MainBlock extends React.Component {
    render() {
        if (this.props.data !== null) { // we know current geolocation
            let MINUTES_TO_BECOME_STALE = 30.0;
            let millisecDiff = Date.now() - (this.props.data.timeLoaded ? this.props.data.timeLoaded : new Date(0));
            let isOld = (millisecDiff / (1000 * 60)) > MINUTES_TO_BECOME_STALE;
            let oldCashedData = isOld ?
                <div className="error-message">
                    Please pay attention you see cashed stale data of your location. In order to update your geolocation or refresh cashed weather data, click the refresh button
                </div> : "";

            return (
                <div>
                    <div className="row main">
                        <MainPreview data={this.props.data}/>
                        <div className="width45">
                            <Forecast data={this.props.data}/>
                        </div>
                    </div>
                    {oldCashedData}
                </div>
            );
        } else {
            this.props.dispatch(addCityByGeolocation());
            return <Loading />;
        }
    }
}

export const MainBlockContainer = connect(
    state => ({ data: state.geolocation })
)(MainBlock);

export default MainBlockContainer;
