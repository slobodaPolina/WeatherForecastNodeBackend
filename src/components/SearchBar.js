import React from 'react';
import { addCityByName, getTheFavorites, addCityToTheFavorites } from '../cityAdder';

export class SearchBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: "",
            isCityValid: true,
            favorites: []
        };
        this.handleSearch = this.handleSearch.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        getTheFavorites(
            (data) => {this.state.favorites.push(data)},
            () => console.error("Problems getting favorite cities!")
        );
    }

    render() {
        let msg = this.state.isCityValid ? "" : <div className="error-message">Seems like the city is not valid</div>;
        return (
            <div>
                <div className="search-bar row">
                    <div className="bigText">Избранное</div>
                    <div>
                        <form onSubmit={this.handleSearch}>
                            <input id="search-input" type="search"
                                placeholder="Добавить новый город" value={this.state.city} onChange={this.handleInputChange}/>
                            <button className="search-button roundButton">+</button>
                        </form>
                    </div>
                </div>
                {msg}
            </div>
        );
    }

    handleInputChange(event) {
        this.setState({city: event.target.value});
    }

    handleSearch(event) {
        const getCallback = (isSuccess) => () => {
            this.setState({isCityValid: isSuccess});
            // remove loading card anyway
            this.props.dispatch({ //todo
                type: 'REMOVE_FAVORITE',
                cityName: this.state.city
            });
            if (isSuccess) { // means we got data from weather service
                addCityToTheFavorites(
                    this.state.city,
                    () => {
                        this.setState({favorites: this.state.favorites.push(this.state.city)});
                        this.setState({city: ""});
                    },
                    () => {
                        console.error('error during saving favorite city in db!');
                        this.setState({isCityValid: false});
                    }
                );
            }
        };

        event.preventDefault();
        if (this.props.favorites.find(cityName => cityName === this.state.city)) { // if we already have it we wont add it again
            this.setState({isCityValid: false});
        } else {
            // it is just to start the loader
            this.props.dispatch({ //todo deal with loading
                type: 'ADD_FAVORITE',
                city: {name: this.state.city}
            });
            // load real data and display it
            addCityByName(
                this.state.city,
                getCallback(true),
                getCallback(false)
            );
        }
    }
}

export default SearchBar;
