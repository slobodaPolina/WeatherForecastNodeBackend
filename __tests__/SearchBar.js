import React from 'react';
import renderer from 'react-test-renderer';
import { SearchBar } from '../src/components/SearchBar';

it('renders SearchBar correctly', () => {
    const tree = renderer.create(
        <SearchBar/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders SearchBar non valid city correctly', () => {
    const component = renderer.create(
        <SearchBar/>
    );
    component.getInstance().setState({isCityValid: false});
    expect(component.toJSON()).toMatchSnapshot();
});

it('renders SearchBar with city correctly', () => {
    const component = renderer.create(
        <SearchBar/>
    );
    component.getInstance().setState({city: "London"});
    expect(component.toJSON()).toMatchSnapshot();
});
