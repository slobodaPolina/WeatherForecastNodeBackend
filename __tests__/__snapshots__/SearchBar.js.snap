// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SearchBar correctly 1`] = `
<div>
  <div
    className="search-bar row"
  >
    <div
      className="bigText"
    >
      Избранное
    </div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <input
          id="search-input"
          onChange={[Function]}
          placeholder="Добавить новый город"
          type="search"
          value=""
        />
        <button
          className="search-button roundButton"
        >
          +
        </button>
      </form>
    </div>
  </div>
  
</div>
`;

exports[`renders SearchBar non valid city correctly 1`] = `
<div>
  <div
    className="search-bar row"
  >
    <div
      className="bigText"
    >
      Избранное
    </div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <input
          id="search-input"
          onChange={[Function]}
          placeholder="Добавить новый город"
          type="search"
          value=""
        />
        <button
          className="search-button roundButton"
        >
          +
        </button>
      </form>
    </div>
  </div>
  <div
    className="error-message"
  >
    Seems like the city is not valid
  </div>
</div>
`;

exports[`renders SearchBar with city correctly 1`] = `
<div>
  <div
    className="search-bar row"
  >
    <div
      className="bigText"
    >
      Избранное
    </div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <input
          id="search-input"
          onChange={[Function]}
          placeholder="Добавить новый город"
          type="search"
          value="London"
        />
        <button
          className="search-button roundButton"
        >
          +
        </button>
      </form>
    </div>
  </div>
  
</div>
`;
